File: Solution/src/SkipIterator.java
import java.util.*;

public class SkipIterator implements Iterator<Integer> {
    private final Iterator<Integer> iterator;
    private final Map<Integer,Integer> map;
    private  Integer cursor;


    SkipIterator(Iterator<Integer> iterator){
    this.iterator = iterator;
    map = new HashMap<>();
    seekCursor();

    }

    public boolean hasNext(){
    return cursor!=null;
    }

    public Integer next(){
    Integer value = cursor;
    seekCursor();
    return value;
    }

    public void skip(int val){
    if(val==cursor){
      seekCursor();
    }
    else{
        if(!map.containsKey(val)){
            map.put(val,0);
        }
        else{
            map.put(val,map.get(val)+1);
        }
    }
    }

    public void seekCursor(){
    cursor=null;

    while(iterator.hasNext()){
        Integer tmp = iterator.next();
        if(!map.containsKey(tmp)){
            cursor=tmp;
            break;
        }
        else{
            if(map.get(tmp)>0){
                map.put(tmp,map.get(tmp)-1);
            }
            else{
                map.remove(tmp);
            }
        }
    }
    }

    public static void main(String[] args) {
        List<Integer> list = Arrays.asList(2,3,5,6,5,7,5,-1,5,10);
        Iterator iterator = list.iterator();
        SkipIterator skipIterator = new SkipIterator(iterator);
        System.out.println(skipIterator.hasNext());
        System.out.println(skipIterator.next());
        skipIterator.skip(5);
        System.out.println(skipIterator.next());
        System.out.println(skipIterator.next());
        System.out.println(skipIterator.next());
        skipIterator.skip(5);
        skipIterator.skip(5);
        System.out.println(skipIterator.next());
        System.out.println(skipIterator.next());
        System.out.println(skipIterator.next());
        System.out.println(skipIterator.hasNext());





    }
}

File: SkipIterator.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Text;

namespace Design
{
    //C# doesnot support inbuilt iterator
    public  class SkipIterator : Iterator<int>
    {
        private Iterator<int> it;
        Dictionary<int, int> map;
        int nextEl;

        public SkipIterator(Iterator<int> it)
        {
            this.map = new Dictionary<int, int>();
            this.it = it;
            advance();
        }
        public SkipIterator() { }

        public bool HasNext()
        {
            return nextEl != Int32.MinValue;
        }

        public int Next()
        {
            int temp = nextEl;
            advance();
            return temp;
        }

        public void advance()
        {
            nextEl = Int32.MinValue;
            while (nextEl == Int32.MinValue && it.ToString() != "") //it.HasNext())
            {
                int el = 1; //it.Next();
                if (!map.ContainsKey(el))
                {
                    nextEl = el;
                }
                else
                {
                    map.Add(el, map[el] - 1);
                    if(map[el] == 0)
                       map.Remove(el);
                }
            }
        }
        public void Skip(int val)
        {
            if (val == nextEl)
            {
                advance();
            }
            else
            {
                map.Add(val, map.GetValueOrDefault(val, 0) + 1);
            }
        }

        public static void Test()
        {
            ArrayList arlist = new ArrayList(){5, 6, 7, 7, 7, 8, 9, 10, 11};
            //Iterator<int> inp = new Iterator<int> { 5, 6, 7, 7, 7, 8, 9, 10, 11 };
            SkipIterator it = new SkipIterator();
            Console.WriteLine(it.Next());
            it.Skip(7);
            it.Skip(7);
            it.Skip(7);
            Console.WriteLine(it.Next());
            Console.WriteLine(it.HasNext());
            Console.WriteLine(it.Next());
            Console.WriteLine(it.Next());
        }
    }

    public class Iterator<T> : IEnumerable<T>
    {
        IEnumerator<T> IEnumerable<T>.GetEnumerator()
        {
            throw new NotImplementedException();
        }

        IEnumerator IEnumerable.GetEnumerator()
        {
            throw new NotImplementedException();
        }
    }
}
